File: Examples/test-suite/python/cpp11_shared_ptr_overload_runme.py
import cpp11_shared_ptr_overload
from cpp11_shared_ptr_overload import MyType

# ref
ret = cpp11_shared_ptr_overload.UseA(MyType("123"))
if ret != "123 ref": raise RuntimeError("UseA fail:" + ret)

ret = cpp11_shared_ptr_overload.UseB(0, MyType("123"))
if ret != "123 ref": raise RuntimeError("UseB fail:" + ret)

ret = cpp11_shared_ptr_overload.UseC(0, MyType("123"), MyType("456"))
if ret != "123 ref": raise RuntimeError("UseC fail:" + ret)

# sharedptr
ret = cpp11_shared_ptr_overload.UseX(MyType("123"))
if ret != "123 sharedptr": raise RuntimeError("UseX fail:" + ret)

ret = cpp11_shared_ptr_overload.UseY(0, MyType("123"))
if ret != "123 sharedptr": raise RuntimeError("UseY fail:" + ret)

ret = cpp11_shared_ptr_overload.UseZ(0, MyType("123"), MyType("456"))
if ret != "123 sharedptr": raise RuntimeError("UseZ fail:" + ret)

# Combo1-4
ret = cpp11_shared_ptr_overload.Combo1(MyType("XXX"))
if ret != "XXXCombo1": raise RuntimeError("Combo1 fail:" + ret)

ret = cpp11_shared_ptr_overload.Combo2(MyType("XXX"))
if ret != "XXXCombo2": raise RuntimeError("Combo2 fail:" + ret)

ret = cpp11_shared_ptr_overload.Combo3(MyType("XXX"))
if ret != "XXXCombo3": raise RuntimeError("Combo3 fail:" + ret)

ret = cpp11_shared_ptr_overload.Combo4(MyType("XXX"))
if ret != "XXXCombo4": raise RuntimeError("Combo4 fail:" + ret)

# Combo5-7
ret = cpp11_shared_ptr_overload.Combo5(MyType("XXX"))
if ret != "XXXCombo5": raise RuntimeError("Combo5 fail:" + ret)

ret = cpp11_shared_ptr_overload.Combo6(MyType("XXX"))
if ret != "XXXCombo6": raise RuntimeError("Combo6 fail:" + ret)

ret = cpp11_shared_ptr_overload.Combo7(MyType("XXX"))
if ret != "XXXCombo7": raise RuntimeError("Combo7 fail:" + ret)

File: Examples/test-suite/lua/cpp11_shared_ptr_overload_runme.lua
require("cpp11_shared_ptr_overload")

local s = cpp11_shared_ptr_overload

-- ref
local ret = s.UseA(s.MyType("123"))
assert(ret == "123 ref", "UseA fail:" .. ret)

ret = s.UseB(0, s.MyType("123"))
assert(ret == "123 ref", "UseB fail:" .. ret)

ret = s.UseC(0, s.MyType("123"), s.MyType("456"))
assert(ret == "123 ref", "UseC fail:" .. ret)

-- sharedptr
ret = s.UseX(s.MyType("123"))
assert(ret == "123 sharedptr", "UseX fail:" .. ret)

ret = s.UseY(0, s.MyType("123"))
assert(ret == "123 sharedptr", "UseY fail:" .. ret)

ret = s.UseZ(0, s.MyType("123"), s.MyType("456"))
assert(ret == "123 sharedptr", "UseZ fail:" .. ret)

-- Combo1-4
ret = s.Combo1(s.MyType("XXX"))
assert(ret == "XXXCombo1", "Combo1 fail:" .. ret)

ret = s.Combo2(s.MyType("XXX"))
assert(ret == "XXXCombo2", "Combo2 fail:" .. ret)

ret = s.Combo3(s.MyType("XXX"))
assert(ret == "XXXCombo3", "Combo3 fail:" .. ret)

ret = s.Combo4(s.MyType("XXX"))
assert(ret == "XXXCombo4", "Combo4 fail:" .. ret)

-- Combo5-7
ret = s.Combo5(s.MyType("XXX"))
assert(ret == "XXXCombo5", "Combo5 fail:" .. ret)

ret = s.Combo6(s.MyType("XXX"))
assert(ret == "XXXCombo6", "Combo6 fail:" .. ret)

ret = s.Combo7(s.MyType("XXX"))
assert(ret == "XXXCombo7", "Combo7 fail:" .. ret)

-- Test default and nil for NULL
assert(s.default_0(), "default_0() fail")
assert(s.default_0(nil), "default_0(nil) fail")
assert(s.default_0l(), "default_0l() fail")
assert(s.default_0l(nil), "default_0l(nil) fail")
assert(s.default_null(), "default_null() fail")
assert(s.default_null(nil), "default_null(nil) fail")
assert(s.default_nullptr(), "default_nullptr() fail")
assert(s.default_nullptr(nil), "default_nullptr(nil) fail")
